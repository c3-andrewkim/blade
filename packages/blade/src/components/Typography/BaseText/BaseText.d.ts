import type { ReactNode } from 'react';
import type { ColorContrast } from '../../../tokens/theme/theme';
import type { Theme } from '../../BladeProvider';

type DotNotationStringToken<TokenType> = {
  [K in keyof TokenType]: `${K & string}.${TokenType[K] extends Record<
    string,
    ColorContrast | string
  >
    ? Extract<keyof TokenType[K], number | string>
    : DotNotationStringToken<TokenType[K]>}`;
}[keyof TokenType];

type FeedbackColors = `feedback.text.${DotNotationStringToken<
  Theme['colors']['feedback']['text']
>}`;
type SurfaceColors = `surface.text.${DotNotationStringToken<Theme['colors']['surface']['text']>}`;
type ActionColors = `action.text.${DotNotationStringToken<Theme['colors']['action']['text']>}`;

export type BaseTextProps = {
  color: ActionColors | FeedbackColors | SurfaceColors;
  fontFamily: keyof Theme['typography']['fonts']['family'];
  fontSize: keyof Theme['typography']['fonts']['size'];
  fontWeight: keyof Theme['typography']['fonts']['weight'];
  fontStyle: 'italic' | 'normal';
  textDecorationLine: 'line-through' | 'none';
  lineHeight: keyof Theme['typography']['lineHeights'];
  as?: 'code' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'span';
  textAlign?: 'center' | 'justify' | 'left' | 'right';
  name?: string;
  children?: ReactNode;
  theme: Theme;
};
