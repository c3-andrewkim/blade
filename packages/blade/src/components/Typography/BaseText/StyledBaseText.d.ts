export { default } from './StyledBaseText.web';
