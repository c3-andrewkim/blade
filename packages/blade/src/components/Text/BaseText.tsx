import { ReactNode, ReactElement } from 'react';
import styled from 'styled-components';
import { ColorContrast } from '../../tokens/theme/theme';
import getIn from '../../utils/getIn';
import { useTheme } from '../BladeProvider';
import type { Theme } from '../BladeProvider';
import isEmpty from '../../utils/isEmpty';

/**
 * - type of typography text to be the value of text/code - done
 * - create types for BaseText - done
 * - make autocomplete work with TS - done
 * - write helper to resolve color - done
 * - filter unwanted dom attrs - done
 * - create function to add data-* elements - deferred
 * - write helper to resolve text - deferred
 * - Organise the storybook
 * - write text component for React native
 */

type DotNotationStringToken<TokenType> = {
  [K in keyof TokenType]: `${K & string}.${TokenType[K] extends Record<
    string,
    ColorContrast | string
  >
    ? Extract<keyof TokenType[K], number | string>
    : DotNotationStringToken<TokenType[K]>}`;
}[keyof TokenType];

type FeedbackColors = `feedback.text.${DotNotationStringToken<
  Theme['colors']['feedback']['text']
>}`;
type SurfaceColors = `surface.text.${DotNotationStringToken<Theme['colors']['surface']['text']>}`;
type ActionColors = `action.text.${DotNotationStringToken<Theme['colors']['action']['text']>}`;

export type BaseTextProps = {
  color: ActionColors | FeedbackColors | SurfaceColors;
  fontFamily: keyof Theme['typography']['fonts']['family'];
  fontSize: keyof Theme['typography']['fonts']['size'];
  fontWeight: keyof Theme['typography']['fonts']['weight'];
  fontStyle: 'italic' | 'normal';
  textDecorationLine: 'line-through' | 'none';
  lineHeight: keyof Theme['typography']['lineHeights'];
  as: 'code' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'span';
  name?: string;
  children?: ReactNode;
};

const StyledBaseText = styled.div<BaseTextProps>(
  ({
    theme,
    color,
    fontFamily,
    fontSize,
    fontWeight,
    fontStyle,
    textDecorationLine,
    lineHeight,
  }) => `
  color: ${color};
  font-family: ${theme.typography.fonts.family[fontFamily]};
  font-size: ${theme.typography.fonts.size[fontSize]}rem;
  font-weight: ${theme.typography.fonts.weight[fontWeight]};
  font-style: ${fontStyle};
  text-decoration-line: ${textDecorationLine};
  line-height: ${lineHeight};
`,
);

const BaseText = ({
  color,
  fontFamily,
  fontSize,
  fontWeight,
  fontStyle,
  textDecorationLine,
  lineHeight,
  as,
  name,
  children,
}: BaseTextProps): ReactElement => {
  const { theme: bladeTheme } = useTheme();
  const textColor = getIn(bladeTheme.colors, color);

  if (isEmpty(textColor)) {
    throw new Error(
      // @todo: generate the error message with valid values
      `[Blade:BaseText]: Invalid color prop ${color} passed, expected value to be one of theme color value from surface.text, feedback.text or action.text`,
    );
  }

  return (
    <StyledBaseText
      data-blade-component={name}
      theme={bladeTheme}
      color={textColor}
      fontFamily={fontFamily}
      fontSize={fontSize}
      fontWeight={fontWeight}
      fontStyle={fontStyle}
      textDecorationLine={textDecorationLine}
      lineHeight={lineHeight}
      as={as}
    >
      {children}
    </StyledBaseText>
  );
};

export default BaseText;
