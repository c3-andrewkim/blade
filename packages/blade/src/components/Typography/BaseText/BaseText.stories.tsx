import type { ComponentStory, Meta } from '@storybook/react';
import type { BaseTextProps } from './BaseText';
import BaseTextComponent from './BaseText';

export default {
  title: 'Components/Typography/BaseText',
  component: BaseTextComponent,
  args: {
    color: 'surface.text.normal.lowContrast',
    fontFamily: 'text',
    fontSize: 700,
    fontWeight: 'regular',
    fontStyle: 'normal',
    textAlign: 'left',
    textDecorationLine: 'none',
    lineHeight: '5xl',
    name: 'Storybook',
    children: 'Base Text',
  },
} as Meta<BaseTextProps>;

const BaseTextTemplate: ComponentStory<typeof BaseTextComponent> = (args) => {
  return <BaseTextComponent {...args}>{args.children}</BaseTextComponent>;
};

export const BaseText = BaseTextTemplate.bind({});
// Need to do this because of storybook's weird naming convention, More details here: https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#single-story-hoisting
BaseText.storyName = 'BaseText';
