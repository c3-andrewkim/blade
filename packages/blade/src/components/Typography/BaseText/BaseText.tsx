import type { ReactElement } from 'react';
import getIn from '../../../utils/getIn';
import { useTheme } from '../../BladeProvider';
import isEmpty from '../../../utils/isEmpty';
import StyledBaseText from './StyledBaseText';
import type { BaseTextProps } from './BaseText.d';

const BaseText = ({
  color,
  fontFamily,
  fontSize,
  fontWeight,
  fontStyle,
  textDecorationLine,
  lineHeight,
  as,
  textAlign,
  ...rest
}: BaseTextProps): ReactElement => {
  const { theme } = useTheme();
  const textColor = getIn(theme.colors, color);

  if (isEmpty(textColor)) {
    throw new Error(`[Blade:BaseText]: Invalid value ${color} for color prop passed`);
  }

  return (
    <StyledBaseText
      color={textColor}
      fontFamily={fontFamily}
      fontSize={fontSize}
      fontWeight={fontWeight}
      fontStyle={fontStyle}
      textDecorationLine={textDecorationLine}
      lineHeight={lineHeight}
      as={as}
      textAlign={textAlign}
    >
      {rest.children}
    </StyledBaseText>
  );
};

export default BaseText;
