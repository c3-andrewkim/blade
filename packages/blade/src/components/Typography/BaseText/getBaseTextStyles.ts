import makeTypography from '../../../utils/makeTypography';
import type { Theme } from '../../BladeProvider';
import type { BaseTextProps } from './BaseText.d';

const getBaseTextStyles = ({
  theme,
  color,
  fontFamily,
  fontSize,
  fontWeight,
  fontStyle = 'normal',
  textDecorationLine = 'none',
  lineHeight,
  textAlign,
}: BaseTextProps & { theme: Theme }): string => `
  color: ${color};
  font-family: ${theme.typography.fonts.family[fontFamily]};
  font-size: ${makeTypography(theme.typography.fonts.size[fontSize])};
  font-weight: ${theme.typography.fonts.weight[fontWeight]};
  font-style: ${fontStyle};
  text-decoration-line: ${textDecorationLine};
  line-height: ${makeTypography(theme.typography.lineHeights[lineHeight])};
  ${textAlign ? `text-align: ${textAlign};` : ''}
`;

export default getBaseTextStyles;
