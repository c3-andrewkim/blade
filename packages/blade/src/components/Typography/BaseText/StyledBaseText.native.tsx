import styled from 'styled-components/native';
import getBaseTextStyles from './getBaseTextStyles';
import type { StyledBaseTextProps } from './StyledBaseText.d';

const StyledBaseText = styled.Text<StyledBaseTextProps>(
  ({
    color,
    fontFamily,
    fontSize,
    fontWeight,
    fontStyle,
    textDecorationLine,
    lineHeight,
    textAlign,
    as,
  }) => {
    if (as) {
      throw new Error(`[Blade: BaseText]: "as" prop is not supported for BaseText on React Native`);
    } else {
      return getBaseTextStyles({
        color,
        fontFamily,
        fontSize,
        fontWeight,
        fontStyle,
        textDecorationLine,
        lineHeight,
        textAlign,
      });
    }
  },
);

export default StyledBaseText;
