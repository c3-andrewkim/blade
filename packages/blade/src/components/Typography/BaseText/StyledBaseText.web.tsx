import styled from 'styled-components';
import getBaseTextStyles from './getBaseTextStyles';
import type { BaseTextProps } from './BaseText.d';

const StyledBaseText = styled.div<BaseTextProps>(
  ({
    theme,
    color,
    fontFamily,
    fontSize,
    fontWeight,
    fontStyle,
    textDecorationLine,
    lineHeight,
    textAlign,
  }) =>
    getBaseTextStyles({
      theme,
      color,
      fontFamily,
      fontSize,
      fontWeight,
      fontStyle,
      textDecorationLine,
      lineHeight,
      textAlign,
    }),
);

export default StyledBaseText;
