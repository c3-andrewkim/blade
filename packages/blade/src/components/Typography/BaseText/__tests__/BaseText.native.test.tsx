import React from 'react';
import renderWithTheme from '../../../../_helpers/testing/renderWithTheme.native';
import BaseText from '../';
import { ERROR_AS_PROP_NOT_SUPPORTED } from '../StyledBaseText.native';
import { getInvalidPropValueError } from '../BaseText';

beforeAll(() => jest.spyOn(console, 'error').mockImplementation());
afterAll(() => jest.restoreAllMocks());

describe('<BaseText />', () => {
  it('should render regular text', () => {
    const displayText = 'Displaying some text';
    const renderedTree = renderWithTheme(
      <BaseText
        color="surface.text.normal.highContrast"
        fontFamily="text"
        lineHeight="m"
        fontSize={25}
        fontWeight="regular"
      >
        {displayText}
      </BaseText>,
    ).toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  it('should render text in italics style with line-through', () => {
    const displayText = 'Displaying some text';
    const renderedTree = renderWithTheme(
      <BaseText
        color="surface.text.normal.highContrast"
        fontFamily="text"
        lineHeight="m"
        fontSize={25}
        fontWeight="regular"
        fontStyle="italic"
        textDecorationLine="line-through"
      >
        {displayText}
      </BaseText>,
    ).toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  it('should throw error when "as" prop passed', () => {
    try {
      const displayText = 'Displaying some text';
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
          as="p"
        >
          {displayText}
        </BaseText>,
      ).toJSON();
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(ERROR_AS_PROP_NOT_SUPPORTED);
      }
    }
  });

  it('should fail when incorrect color passed', () => {
    const displayText = 'Displaying some text';
    const color = 'somerandomcolor';
    try {
      renderWithTheme(
        <BaseText
          // @ts-expect-error testing the invalid value for color prop error case
          color={color}
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      ).toJSON();
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'color', propValue: color }),
        );
      }
    }
  });
  it('should fail when incorrect fontFamily passed', () => {
    const displayText = 'Displaying some text';
    const fontFamily = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          // @ts-expect-error testing the invalid value for fontFamily prop error case
          fontFamily={fontFamily}
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'fontFamily', propValue: fontFamily }),
        );
      }
    }
  });

  it('should fail when incorrect fontSize passed', () => {
    const displayText = 'Displaying some text';
    const fontSize = 1588;
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          // @ts-expect-error testing the invalid value for fontSize prop error case
          fontSize={1588}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'fontSize', propValue: fontSize }),
        );
      }
    }
  });

  it('should fail when incorrect fontWeight passed', () => {
    const displayText = 'Displaying some text';
    const fontWeight = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          // @ts-expect-error testing the invalid value for fontWeight prop error case
          fontWeight={fontWeight}
          fontStyle="italic"
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'fontWeight', propValue: fontWeight }),
        );
      }
    }
  });

  it('should fail when incorrect fontStyle passed', () => {
    const displayText = 'Displaying some text';
    const fontStyle = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          // @ts-expect-error testing the invalid value for fontStyle prop error case
          fontStyle={fontStyle}
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'fontStyle', propValue: fontStyle }),
        );
      }
    }
  });

  it('should fail when incorrect textDecorationLine passed', () => {
    const displayText = 'Displaying some text';
    const textDecorationLine = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          // @ts-expect-error testing the invalid value for textDecorationLine prop error case
          textDecorationLine={textDecorationLine}
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({
            propName: 'textDecorationLine',
            propValue: textDecorationLine,
          }),
        );
      }
    }
  });

  it('should fail when incorrect lineHeight passed', () => {
    const displayText = 'Displaying some text';
    const lineHeight = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          // @ts-expect-error testing the invalid value for lineHeight prop error case
          lineHeight={lineHeight}
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'lineHeight', propValue: lineHeight }),
        );
      }
    }
  });

  it('should fail when incorrect textAlign passed', () => {
    const displayText = 'Displaying some text';
    const textAlign = 'somerandom';
    try {
      renderWithTheme(
        <BaseText
          color="surface.text.normal.highContrast"
          fontFamily="text"
          lineHeight="m"
          fontSize={25}
          fontWeight="regular"
          fontStyle="italic"
          textDecorationLine="line-through"
          // @ts-expect-error testing the invalid value for textAlign prop error case
          textAlign={textAlign}
        >
          {displayText}
        </BaseText>,
      );
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toEqual(
          getInvalidPropValueError({ propName: 'textAlign', propValue: textAlign }),
        );
      }
    }
  });
});
