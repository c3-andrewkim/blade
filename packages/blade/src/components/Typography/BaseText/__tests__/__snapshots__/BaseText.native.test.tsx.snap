// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseText /> should render regular text 1`] = `
<Text
  color="hsla(0, 0%, 100%, 1)"
  fontFamily="text"
  fontSize={25}
  fontWeight="regular"
  lineHeight="m"
  style={
    Array [
      Object {
        "color": "hsla(0, 0%, 100%, 1)",
        "fontFamily": "Lato",
        "fontSize": 11,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 16,
        "textDecorationLine": "none",
      },
    ]
  }
>
  Displaying some text
</Text>
`;

exports[`<BaseText /> should render text in italics style with line-through 1`] = `
<Text
  color="hsla(0, 0%, 100%, 1)"
  fontFamily="text"
  fontSize={25}
  fontStyle="italic"
  fontWeight="regular"
  lineHeight="m"
  style={
    Array [
      Object {
        "color": "hsla(0, 0%, 100%, 1)",
        "fontFamily": "Lato",
        "fontSize": 11,
        "fontStyle": "italic",
        "fontWeight": "400",
        "lineHeight": 16,
        "textDecorationLine": "line-through",
      },
    ]
  }
  textDecorationLine="line-through"
>
  Displaying some text
</Text>
`;

exports[`<BaseText /> should throw error when "as" prop passed 1`] = `
<p
  color="hsla(0, 0%, 100%, 1)"
  fontFamily="text"
  fontSize={25}
  fontStyle="italic"
  fontWeight="regular"
  lineHeight="m"
  style={
    Array [
      Object {
        "color": "hsla(0, 0%, 100%, 1)",
        "fontFamily": "Lato",
        "fontSize": 11,
        "fontStyle": "italic",
        "fontWeight": "400",
        "lineHeight": 16,
        "textDecorationLine": "line-through",
      },
    ]
  }
  textDecorationLine="line-through"
>
  Displaying some text
</p>
`;
